### PR TITLE
Use a cached prism-go-client with session auth

### PR DESCRIPTION
**Pull request by thunderboltsid**
This will reduce the number of basic auth calls that will be made to IAM stack.

### CODE REVIEW

File: internal/testing/mock/mock_client.go [hunk 1]
```diff
@@ -38,10 +37,12 @@ func CreateMockClient(mockEnvironment MockEnvironment) *MockClient {
 	}
 }
 
+// Get returns the mockPrism
 func (mc *MockClient) Get() (interfaces.Prism, error) {
 	return &mc.mockPrism, nil
 }
 
+// SetInformers sets the sharedInformers
```
Comment (thunderboltsid) [inline internal/testing/mock/mock_client.go:45]: This file had to be modified due to failing lint errors.

**Review by adiantum — APPROVED** [review]

File: pkg/provider/client.go [hunk 2]
```diff
@@ -20,64 +20,66 @@ import (
...
 	if err != nil {
-		return nil, err
-	}
-	creds := &prismgoclient.Credentials{
-		URL:      me.Address.Host, // Not really an URL
-		Endpoint: me.Address.Host,
-		Insecure: me.Insecure,
-		Username: me.ApiCredentials.Username,
-		Password: me.ApiCredentials.Password,
+		klog.Errorf("failed to get management endpoint: %v", err)
```
Comment (deepakm-ntnx) [inline pkg/provider/client.go:65]: nit: %s with err.Error() should suffice

File: pkg/provider/client.go [hunk 3]
```diff
@@ -20,64 +20,66 @@ import (
...
-	}
+	return *mgmtEndpoint
+}
 
-	nutanixClient, err := prismClientV3.NewV3Client(*creds, clientOpts...)
-	if err != nil {
-		return nil, err
+func (n *nutanixClient) Get() (interfaces.Prism, error) {
+	if err := n.setupEnvironment(); err != nil {
+		return nil, fmt.Errorf("%s: %v", errEnvironmentNotReady, err)
```
Comment (deepakm-ntnx) [inline pkg/provider/client.go:74]: nit: %s with err.Error() should suffice

Comment (thunderboltsid) [reply pkg/provider/client.go]: since it's a fmt.Errorf clause, I've switched to `%w`  to wrap the error

File: pkg/provider/client.go (hunk 2 above)
Comment (thunderboltsid) [reply pkg/provider/client.go]: good point. fixed.

**Review by deepakm-ntnx — APPROVED** [review]
there are many codecov annotations in pkg/provider/client.go. are you planning to address them in this PR?

File: pkg/provider/client.go [hunk 4]
```diff
@@ -20,64 +20,66 @@ import (
...
-	nutanixClient, err := prismClientV3.NewV3Client(*creds, clientOpts...)
-	if err != nil {
-		return nil, err
+func (n *nutanixClient) Get() (interfaces.Prism, error) {
+	if err := n.setupEnvironment(); err != nil {
+		return nil, fmt.Errorf("%s: %w", errEnvironmentNotReady, err)
 	}
 
-	_, err = nutanixClient.V3.GetCurrentLoggedInUser(context.Background())
+	client, err := n.clientCache.GetOrCreate(n)
```
Comment (deepakm-ntnx) [inline pkg/provider/client.go:77]: I think we wont be deleting it from the cache. also there will always be only one instance in the cache hopefully given single cluster

Comment (thunderboltsid) [reply pkg/provider/client.go:77]: correct. this is to just offload client creation to an existing method in the client.

**Pull request merged**